src/search/move_picker.h

```diff
@@ -252,68 +237,54 @@ class MovePicker {
         return bestMove;
     }
 
-    constexpr std::optional<movegen::Move> pickKillerMove(KillerMoveType type)
+    // TODO template over bool hasCounter, to avoid redundant check?
+    template<Player player>
+    void generateQuietScores(const BitBoard& board)
     {
         const auto killerMoves = m_searchTables.getKillerMove(m_ply);
 
-        const auto killerMove = type == KillerMoveType::First ? killerMoves.first : killerMoves.second;
-
         for (uint16_t i = 0; i < m_moves.count(); i++) {
+            if (!m_moves[i].isQuietMove() || m_moves[i].isNull())
+                continue;
+
+            if (m_moves[i] == killerMoves.first) {
+                m_scores[i] = QuietMoveOffsets::KillerMoveFirst;
+            } else if (m_moves[i] == killerMoves.second) {
+                m_scores[i] = QuietMoveOffsets::KillerMoveSecond;
+            } else if (m_prevMove && m_moves[i] == m_searchTables.getCounterMove(m_prevMove.value())) {
+                m_scores[i] = QuietMoveOffsets::CounterMove;
+            } else {
+                const auto attacker = board.getAttackerAtSquare<player>(m_moves[i].fromSquare());
+                m_scores[i] = m_searchTables.getHistoryMove(attacker.value(), m_moves[i].toPos());
 
-            if (!m_moves[i].isNull() && m_moves[i] == killerMove && m_moves[i].isQuietMove()) {
-                auto pickedMove = m_moves[i];
-
-                m_moves.nullifyMove(i);
-
-                return pickedMove;
-            }
-        }
-        return std::nullopt;
-    }
-
-    constexpr std::optional<movegen::Move> pickCounterMove()
-    {
-        if (!m_prevMove.has_value())
-            return std::nullopt;
-
-        const auto counterMove = m_searchTables.getCounterMove(m_prevMove.value());
-
-        for (uint16_t i = 0; i < m_moves.count(); i++) {
-            if (!m_moves[i].isNull() && m_moves[i].isQuietMove() && m_moves[i] == counterMove) {
-                auto pickedMove = m_moves[i];
-
-                m_moves.nullifyMove(i);
-
-                return pickedMove;
+                // Debug
+                if (m_scores[i] > QuietMoveOffsets::CounterMove)
+                    assert(false);
```

Comment:
`assert` is supposed to be used like this:
```suggestion
                assert(m_scores[i] < QuietMoveOffsets::CounterMove);
```

then when compiling with `NDEBUG` the entire macro results in `void(0)`

Reply:
That's really hepful to know know about `NDEBUG` - thanks!

On reflection, I don't think this statement should be included in either debug or release builds - because it doesn't correspond to a genuine bug, only sub-optimal choices of move ordering. I'll remove it.